lib/src/iterable_extensions.dart

```diff
@@ -0,0 +1,519 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Extension methods on iterables.
+
+import 'package:collection/src/utils.dart';
+
+import "algorithms.dart";
+
+/// Extensions that apply to all iterables.
+extension IterableExtension<T> on Iterable<T> {
+
+  /// The elements that do not satisfy [test].
+  Iterable<T> whereNot(bool test(T element)) =>
+      where((element) => !test(element));
+
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the [compare] [Comparator].
+  List<T> sorted(int compare(T a, T b)) => [...this]..sort(compare);
+
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the [compare] [Comparator] of the
+  /// property [keyOf] of the element.
+  List<T> sortedCompareBy<K>(K keyOf(T element), int compare(K a, K b),
+      [int start = 0, int end]) {
+    var elements = [...this];
+    quickSortBy<T, K>(elements, keyOf, compare, start, end);
+    return elements;
+  }
+
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// The elements are ordered by the natural ordering of the
+  /// property [keyOf] of the element.
+  List<T> sortedBy<K extends Comparable<K>>(K keyOf(T element),
+      [int start = 0, int end]) {
+    var elements = [...this];
+    quickSortBy<T, K>(elements, keyOf, defaultCompare<K>(), start, end);
+    return elements;
+  }
+
+  /// Whether the elements are sorted by the [compare] ordering.
+  bool isSorted(int compare(T a, T b)) {
+    var iterator = this.iterator;
+    if (!iterator.moveNext()) return true;
+    var previousElement = iterator.current;
+    while (iterator.moveNext()) {
+      var element = iterator.current;
+      if (compare(previousElement, element) > 0) return false;
+      previousElement = element;
+    }
+    return true;
+  }
+
+  /// Whether the elements are sorted by the [compare] ordering of [keyOf].
+  bool isSortedBy<K>(K keyOf(T element), int compare(K a, K b)) {
+    var iterator = this.iterator;
+    if (!iterator.moveNext()) return true;
+    var previousKey = keyOf(iterator.current);
+    while (iterator.moveNext()) {
+      var key = keyOf(iterator.current);
+      if (compare(previousKey, key) > 0) return false;
+      previousKey = key;
+    }
+    return true;
+  }
+
+  /// The elements of this iterable with duplicates removed.
+  Iterable<T> unique() sync* {
+    var seen = <T>{};
```

Comment:
Would `HashSet<T>();` be cheaper? We don't care about maintaining insertion order.